oct/drivers.py

```diff
@@ -17,9 +22,36 @@ def get_driver(browser: str, grid: str) -> WebDriver:
         options.add_argument("--ignore-certificate-errors")
         options.add_argument("--remote-debugging-port=9222")
         options.headless = True
-        output = webdriver.Remote(
-            command_executor=grid, desired_capabilities=DesiredCapabilities.CHROME, options=options
-        )
+
+        # poll server connection
+        try:
+            url = grid.replace('/wd/hub', '')
+            polling2.poll(
+                lambda: requests.get(url).status_code == 200,
+                ignore_exceptions=(ConnectionError,),
+                step=1,
+                timeout=10
+            )
+        except polling2.TimeoutException:
+            raise ConnectionError("Could not connect to Selenium server")
+
+        # poll driver initialization
+        try:
+            output = polling2.poll(
+                lambda: webdriver.Remote(
+                    command_executor=grid,
+                    desired_capabilities=DesiredCapabilities.CHROME,
+                    options=options
+                ),
+                ignore_exceptions=(WebDriverException,),
+                step=1,
+                timeout=10
+            )
+        except polling2.TimeoutException:
+            raise WebDriverException("Could not start new Driver session")
+        # output = webdriver.Remote(
+        #     command_executor=grid, desired_capabilities=DesiredCapabilities.CHROME, options=options
+        # )
```

Comment:
Not quite I was talking about, but conceptually same

Reply:
because of the strange nature of the selenium node-hub interaction (more of a java stuff) simple :4444 request polling was not enough. Even though the connection to selenium server was established webdriver was failing to instantiate. So additional polling for driver initialization was added